### PR TITLE
[fast-client] Long tail retry budget for fast client

### PR DESCRIPTION
**Pull request by xunyin8**
## [fast-client] Long tail retry budget for fast client
1. Introduced RetryManager to measure and calculate retry budget based on a percentage of requests.

2. The RetryManager is only applied to long tail retries and currently hard coded to be 10% of the total original request rate. This means if there were any exception that's not a 429 the retry will be triggered without going through the RetryManager. If the retry budget is exhausted then the retry task will do nothing and the request will either complete eventually (original future) or time out.

## How was this PR tested?
Existing and new tests

## Does this PR introduce any user-facing changes?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Make sure to explain your proposed changes and call out the behavior change.

### CODE REVIEW

**Comment by xunyin8** [comment]
Will add the following metrics to track retry budget behavior:

- The long tail retry budget value that's currently being enforced
- Token bucket stats for the long tail retry budget to monitor usage

**Comment by xunyin8** [comment]
Will also add more unit tests in the `RetriableAvroGenericStoreClient` layer and have at least one integration test with a much lower enforcement window for a smoke test (since long tail retry trigger is not deterministic in integration test).

**Comment by xunyin8** [comment]
> Will add the following metrics to track retry budget behavior:
> 
> * The long tail retry budget value that's currently being enforced
> * Token bucket stats for the long tail retry budget to monitor usage

Done

File: clients/venice-client/src/main/java/com/linkedin/venice/fastclient/meta/RetryManager.java [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+
+
+/**
+ * This class offers advanced client retry behaviors. Specifically enforcing a retry budget and relevant monitoring to
+ * avoid retry storm and alert users when the retry threshold is misconfigured or service is degrading.
+ */
+public class RetryManager implements Closeable {
+  private static final Logger LOGGER = LogManager.getLogger(RetryManager.class);
+  private static final int TOKEN_BUCKET_REFILL_INTERVAL_IN_SECONDS = 1;
+  private static final int TOKEN_BUCKET_CAPACITY_MULTIPLE = 5;
```
Comment (gaojieliu) [inline clients/venice-client/src/main/java/com/linkedin/venice/fastclient/meta/RetryManager.java:23]: Can you leave some javadoc about this field to explain the reasoning?

File: clients/venice-client/src/main/java/com/linkedin/venice/fastclient/meta/RetryManager.java [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+    }
+  }
+
+  public boolean isRetryAllowed() {
+    if (!retryBudgetEnabled.get() || retryTokenBucket.get() == null) {
+      // All retries are allowed when the feature is disabled or during the very first enforcement window when we
+      // haven't collected enough data points yet
+      return true;
+    }
+    return retryTokenBucket.get().tryConsume(1);
```
Comment (gaojieliu) [inline clients/venice-client/src/main/java/com/linkedin/venice/fastclient/meta/RetryManager.java:66]: As `retryTokenBucket` can change, so shall we keep a local reference to avoid the potentially race conditions?
```
TokenBucket tokenBucket = retryTokenBucket.get();
if (!retryBudgetEnabled.get() || tokenBucket == null) {
      // All retries are allowed when the feature is disabled or during the very first enforcement window when we
      // haven't collected enough data points yet
      return true;
}
return tokenBucket.tryConsume(1);
```

File: clients/venice-client/src/main/java/com/linkedin/venice/fastclient/meta/RetryManager.java [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+        long requestCountSinceLastUpdate = requestCount.getAndSet(0);
+        lastUpdateTimestamp = clock.millis();
+        // Minimum user request per second will be 1
+        long newQPS = (long) Math
+            .ceil((double) requestCountSinceLastUpdate / (double) TimeUnit.MILLISECONDS.toSeconds(elapsedTimeInMs));
+        if (previousQPS > 0) {
+          long difference = Math.abs(previousQPS - newQPS);
+          double differenceInPercentDecimal = (double) difference / (double) previousQPS;
+          if (differenceInPercentDecimal > 0.1) {
+            // Only update the retry token bucket if the change in request per seconds is more than 10 percent
```
Comment (gaojieliu) [inline clients/venice-client/src/main/java/com/linkedin/venice/fastclient/meta/RetryManager.java:82]: Why can't we update the token bucket all the time?
Is it an expensive operation?

File: clients/venice-client/src/main/java/com/linkedin/venice/fastclient/stats/RetryManagerStats.java [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+      }
+    }, "retry_limit_per_seconds"));
+    retriesRemaining = registerSensor(new AsyncGauge((ignored, ignored2) -> {
+      TokenBucket bucket = retryManager.getRetryTokenBucket();
+      if (bucket == null) {
+        return -1;
+      } else {
+        return bucket.getStaleTokenCount();
+      }
+    }, "retries_remaining"));
```
Comment (gaojieliu) [inline clients/venice-client/src/main/java/com/linkedin/venice/fastclient/stats/RetryManagerStats.java:32]: Can we also have a metric to track denied retry requests?
I hope we can track retry stats per request type.

**Review by gaojieliu — COMMENTED** [review]
Looks good overall, and left a few clarification comments.

File: clients/venice-client/src/main/java/com/linkedin/venice/fastclient/RetriableAvroGenericStoreClient.java [hunk 5]
```diff
@@ -315,6 +345,7 @@ private <R extends MultiKeyRequestContext<K, V>, RESPONSE> void retryStreamingMu
             savedException,
             pendingKeysFuture,
             scheduledRetryTask));
+    longTailRetryManager.recordRequest();
```
[the remark this answers is not in the document]
Comment (xunyin8) [reply clients/venice-client/src/main/java/com/linkedin/venice/fastclient/RetriableAvroGenericStoreClient.java:348]: Currently retry budget is at the request level, perhaps we should maintain it at the key level and use KPS then?

File: clients/venice-client/src/main/java/com/linkedin/venice/fastclient/meta/RetryManager.java (hunk 3 above)
Comment (xunyin8) [reply clients/venice-client/src/main/java/com/linkedin/venice/fastclient/meta/RetryManager.java:82]: It's not an expensive operation. However, the default initial value for token bucket is `capacity` which is maxed out. If we keep initializing new token bucket then we are technically allowing more retry even if the traffic remains the same.

File: clients/venice-client/src/main/java/com/linkedin/venice/fastclient/meta/RetryManager.java (hunk 2 above)
Comment (xunyin8) [reply clients/venice-client/src/main/java/com/linkedin/venice/fastclient/meta/RetryManager.java:66]: Keeping a local reference just changes the behavior but doesn't really avoid any potential race conditions IMO. We are not taking a lock and blocking when consuming and and updating `retryTokenBucket`. We could get the local reference before or after `retryTokenBucket` is updated.

File: clients/venice-client/src/main/java/com/linkedin/venice/fastclient/RetriableAvroGenericStoreClient.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (gaojieliu) [reply clients/venice-client/src/main/java/com/linkedin/venice/fastclient/RetriableAvroGenericStoreClient.java:348]: I think request level is fine, and KPS level will be problematic if not considering request type.
For example, avg key count for multi-get request is 1000 and if the key based retry budget is being utlized by single-get request, it will allow a very high number of single get requests, which will be very bad.

File: clients/venice-client/src/main/java/com/linkedin/venice/fastclient/meta/RetryManager.java (hunk 2 above)
Comment (gaojieliu) [reply clients/venice-client/src/main/java/com/linkedin/venice/fastclient/meta/RetryManager.java:66]: yeah, and this code I mentioned is mostly to avoid the case that first call of `retryTokenBucket.get()` is not `null`, but the second call of `retryTokenBucket.get()` is `null`.

File: clients/venice-client/src/main/java/com/linkedin/venice/fastclient/meta/RetryManager.java (hunk 3 above)
Comment (gaojieliu) [reply clients/venice-client/src/main/java/com/linkedin/venice/fastclient/meta/RetryManager.java:82]: Emm.. 
So the initialization would assume the max cap, and I thought it should be initialized by the 0.1 * previous QPS, no? Maybe not for the first time, but for the subsequent initialization, it doesn't necessary to initialize with max cap, right?

Comment (xunyin8) [reply clients/venice-client/src/main/java/com/linkedin/venice/fastclient/meta/RetryManager.java:82]: I meant the max cap as in the 0.1 * previous QPS * `TOKEN_BUCKET_CAPACITY_MULTIPLE` (5).

File: clients/venice-client/src/main/java/com/linkedin/venice/fastclient/ClientConfig.java [hunk 6]
```diff
@@ -440,6 +452,8 @@ public static class ClientConfigBuilder<K, V, T extends SpecificRecord> {
 
     private boolean projectionFieldValidation = true;
 
+    private long longTailRetryBudgetEnforcementWindowInMs = 60000; // 1 minute
```
[the remark this answers is not in the document]
Comment (xunyin8) [reply clients/venice-client/src/main/java/com/linkedin/venice/fastclient/ClientConfig.java:455]: 1 minute, I think 5 minutes is too long. The traffic pattern can change multiple times within 5 minutes and not able to adjust the retry budget in a timely manner may cause issues.

File: clients/venice-client/src/main/java/com/linkedin/venice/fastclient/RetriableAvroGenericStoreClient.java [hunk 7]
```diff
@@ -230,6 +252,12 @@ public void compute(
         });
   }
 
+  @Override
+  public void close() {
+    super.close();
+    longTailRetryManager.close();
```
[the remark this answers is not in the document]
Comment (xunyin8) [reply clients/venice-client/src/main/java/com/linkedin/venice/fastclient/RetriableAvroGenericStoreClient.java]: Good catch, let me fix that.

File: clients/venice-client/src/main/java/com/linkedin/venice/fastclient/RetriableAvroGenericStoreClient.java [hunk 8]
```diff
@@ -269,24 +297,26 @@ private <R extends MultiKeyRequestContext<K, V>, RESPONSE> void retryStreamingMu
...
-        streamingRequestExecutor.trigger(
-            retryRequestContext,
-            pendingKeys,
-            getStreamingCallback(
-                retryRequestContext,
-                finalRequestCompletionFuture,
-                savedException,
-                pendingKeysFuture,
-                null));
+        if (throwable != null || longTailRetryManager.isRetryAllowed()) {
```
[the remark this answers is not in the document]
Comment (xunyin8) [reply clients/venice-client/src/main/java/com/linkedin/venice/fastclient/RetriableAvroGenericStoreClient.java]: The long tail scenario. If the original request is taking too long then we won't see any errors yet but we want to trigger the long tail retry because it might come back fast (before the original request). 

We always send out the original/first request. If there is no retry budget and the original request never comes back then the whole request will timeout and fail (the final future).

File: clients/venice-client/src/main/java/com/linkedin/venice/fastclient/ClientConfig.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (gaojieliu) [reply clients/venice-client/src/main/java/com/linkedin/venice/fastclient/ClientConfig.java:455]: yes, but the javadoc mentioned something about 5mins.

File: clients/venice-client/src/main/java/com/linkedin/venice/fastclient/meta/RetryManager.java (hunk 3 above)
Comment (gaojieliu) [reply clients/venice-client/src/main/java/com/linkedin/venice/fastclient/meta/RetryManager.java:103]: What is the reason to multiple `TOKEN_BUCKET_CAPACITY_MULTIPLE`? To tolerate some random QPS spike?

File: clients/venice-client/src/main/java/com/linkedin/venice/fastclient/ClientConfig.java [hunk 9]
```diff
@@ -84,6 +84,12 @@ public class ClientConfig<K, V, T extends SpecificRecord> {
    * route requests to the correct server/partition
    */
   private final GrpcClientConfig grpcClientConfig;
+  /**
+   * The time window used to calculate user traffic and corresponding long tail retry budget. The default value is 5
```
Comment (gaojieliu) [inline clients/venice-client/src/main/java/com/linkedin/venice/fastclient/ClientConfig.java:88]: The javadoc mentions 5mins and the following config is 1min:
```
    private long longTailRetryBudgetEnforcementWindowInMs = 60000; // 1 minute
```
Are these two related?

File: clients/venice-client/src/main/java/com/linkedin/venice/fastclient/RetriableAvroGenericStoreClient.java [hunk 10]
```diff
@@ -37,7 +38,18 @@ public class RetriableAvroGenericStoreClient<K, V> extends DelegatingAvroStoreCl
   private final int longTailRetryThresholdForBatchGetInMicroSeconds;
   private final int longTailRetryThresholdForComputeInMicroSeconds;
   private final TimeoutProcessor timeoutProcessor;
+  /**
+   * The long tail retry budget is only applied to long tail retries. If there were any exception that's not a 429 the
+   * retry will be triggered without going through the long tail {@link RetryManager}. If the retry budget is exhausted
+   * then the retry task will do nothing and the request will either complete eventually (original future) or time out.
+   */
+  private RetryManager singleGetLongTailRetryManager = null;
+  private RetryManager multiGetLongTailRetryManager = null;
```
Comment (gaojieliu) [inline clients/venice-client/src/main/java/com/linkedin/venice/fastclient/RetriableAvroGenericStoreClient.java:47]: Do we have a retry manager for compute request type?

Comment (gaojieliu) [reply clients/venice-client/src/main/java/com/linkedin/venice/fastclient/RetriableAvroGenericStoreClient.java:47]: Or we are planning to use the multiget retry manager for compute request type?

File: clients/venice-client/src/main/java/com/linkedin/venice/fastclient/meta/RetryManager.java (hunk 3 above)
Comment (gaojieliu) [reply clients/venice-client/src/main/java/com/linkedin/venice/fastclient/meta/RetryManager.java:103]: Does this mean fast-client will allow 50% of retry all the time or just some short period and the bursty retry will still draw from the total budget, right?

File: clients/venice-client/src/main/java/com/linkedin/venice/fastclient/meta/RetryManager.java [hunk 11]
```diff
@@ -0,0 +1,142 @@
...
+   */
+  private static final int TOKEN_BUCKET_CAPACITY_MULTIPLE = 5;
+  private final AtomicBoolean retryBudgetEnabled = new AtomicBoolean();
+  private final long enforcementWindowInMs;
+  private final double retryBudgetInPercentDecimal;
+  private final AtomicLong requestCount = new AtomicLong();
+  private final AtomicReference<TokenBucket> retryTokenBucket = new AtomicReference<>(null);
+  private final ScheduledExecutorService scheduler;
+  private final Clock clock;
+  RetryManagerStats retryManagerStats;
```
Comment (gaojieliu) [inline clients/venice-client/src/main/java/com/linkedin/venice/fastclient/meta/RetryManager.java:37]: Can this variable be a `private` member?

**Review by gaojieliu — COMMENTED** [review]
Left a few minor comments.

File: clients/venice-client/src/main/java/com/linkedin/venice/fastclient/RetriableAvroGenericStoreClient.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (mynameborat) [reply clients/venice-client/src/main/java/com/linkedin/venice/fastclient/RetriableAvroGenericStoreClient.java]: thanks for the explanation.

File: clients/venice-client/src/main/java/com/linkedin/venice/fastclient/ClientConfig.java (hunk 9 above)
Comment (xunyin8) [reply clients/venice-client/src/main/java/com/linkedin/venice/fastclient/ClientConfig.java:88]: Yes, good catch. FIxed.

File: clients/venice-client/src/main/java/com/linkedin/venice/fastclient/RetriableAvroGenericStoreClient.java (hunk 10 above)
Comment (gaojieliu) [reply clients/venice-client/src/main/java/com/linkedin/venice/fastclient/RetriableAvroGenericStoreClient.java:47]: Yeah, you are correct.

**Review by gaojieliu — APPROVED** [review]
Thanks for the change!
